lib-rt/CPy.h

```diff
@@ -538,14 +538,43 @@ static void CPy_AddTraceback(const char *filename, const char *funcname, int lin
     Py_DECREF(frame_obj);
 }
 
+// mypyc is not very good at dealing with refcount management of
+// pointers that might be NULL. As a workaround for this, the
+// exception APIs that might want to return NULL pointers instead
+// return properly refcounted pointers to this dummy object.
+struct ExcDummyStruct { PyObject_HEAD };
+static struct ExcDummyStruct _CPy_ExcDummyStruct = { PyObject_HEAD_INIT(&PyBaseObject_Type) };
+static PyObject *_CPy_ExcDummy = (PyObject *)&_CPy_ExcDummyStruct;
+
+static inline void _CPy_ToDummy(PyObject **p) {
+    if (*p == NULL) {
+        Py_INCREF(_CPy_ExcDummy);
+        *p = _CPy_ExcDummy;
+    }
+}
+
+static inline PyObject *_CPy_FromDummy(PyObject *p) {
+    if (p == _CPy_ExcDummy) return NULL;
+    Py_INCREF(p);
+    return p;
+}
+
 static void CPy_CatchError(PyObject **p_type, PyObject **p_value, PyObject **p_traceback) {
-    PyErr_Fetch(p_type, p_value, p_traceback);
```

Comment:
It's worth adding some comments here. The exception handling logic is pretty tricky.